### PR TITLE
feat(query): set setting support expr

### LINKED ISSUE #9456: feat: settings support expression
**Summary**

```
set max_memory_usage=1024*1024*1024*2
```

### PR DESCRIPTION
**Pull request by TCeason**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Support sql

```sql
set max_memory_usage = 1024*1024*1024*2
```

Closes #9456

### CODE REVIEW

File: src/query/sql/src/planner/binder/setting.rs [hunk 1]
```diff
@@ -34,30 +42,51 @@ impl<'a> Binder {
...
 
-        let box (value, _) = type_checker.resolve_literal(value, Some(DataType::String))?;
-        let value = String::from_utf8(
-            value
-                .as_string()
-                .ok_or_else(|| ErrorCode::Internal("Invalid variable value"))?
-                .clone(),
-        )?;
+        let box (scalar, data_type) = type_checker.resolve(value, None).await?;
+        let schema = Arc::new(DataSchema::new(vec![DataField::new("result", data_type)]));
```
Comment (sundy-li) [inline src/query/sql/src/planner/binder/setting.rs:57]: DataSchema::empty

**Review by sundy-li — APPROVED** [review]

File: src/query/sql/src/planner/binder/setting.rs [hunk 2]
```diff
@@ -34,30 +41,51 @@ impl<'a> Binder {
...
+        let box (scalar, _) = type_checker.resolve(value, None).await?;
+        let schema = Arc::new(DataSchema::empty());
 
+        let builder = PhysicalScalarBuilder::new(&schema);
+        let scalar = builder.build(&scalar)?;
+        let expr = scalar.as_expr()?;
+        let block = DataBlock::empty();
+        let func_ctx = self.ctx.try_get_function_context()?;
+        let evaluator = Evaluator::new(&block, func_ctx, &BUILTIN_FUNCTIONS);
+        let val = evaluator
```
Comment (b41sh) [inline src/query/sql/src/planner/binder/setting.rs:64]: we can use `eval_function` here.

File: src/query/sql/src/planner/binder/setting.rs [hunk 3]
```diff
@@ -34,30 +41,51 @@ impl<'a> Binder {
...
+            .run(&expr)
+            .map_err(|_| ErrorCode::SemanticError(format!("Failed to run expr: {}", expr)))?;
+        let value = Literal::try_from(val.into_scalar().unwrap())?;
+        let value = match value {
```
Comment (b41sh) [inline src/query/sql/src/planner/binder/setting.rs:68]: add a`to_string` function for the value.

**Review by ClSlaid — APPROVED** [review]

File: src/query/sql/src/planner/binder/setting.rs [hunk 4]
```diff
@@ -34,30 +41,51 @@ impl<'a> Binder {
...
         let variable = variable.name.clone();
 
-        let box (value, _) = type_checker.resolve_literal(value, Some(DataType::String))?;
-        let value = String::from_utf8(
-            value
-                .as_string()
-                .ok_or_else(|| ErrorCode::Internal("Invalid variable value"))?
-                .clone(),
-        )?;
+        let box (scalar, _) = type_checker.resolve(value, None).await?;
```
Comment (andylokandy) [inline src/query/sql/src/planner/binder/setting.rs:55]: We can add common_ast::Expr::Cast here.

File: src/query/sql/src/planner/binder/setting.rs [hunk 5]
```diff
@@ -34,30 +41,51 @@ impl<'a> Binder {
...
+        let builder = PhysicalScalarBuilder::new(&schema);
+        let scalar = builder.build(&scalar)?;
+        let expr = scalar.as_expr()?;
+        let block = DataBlock::empty();
+        let func_ctx = self.ctx.try_get_function_context()?;
+        let evaluator = Evaluator::new(&block, func_ctx, &BUILTIN_FUNCTIONS);
+        let val = evaluator
+            .run(&expr)
+            .map_err(|_| ErrorCode::SemanticError(format!("Failed to run expr: {}", expr)))?;
+        let value = Literal::try_from(val.into_scalar().unwrap())?;
```
Comment (andylokandy) [inline src/query/sql/src/planner/binder/setting.rs:67]: it may be column therefore will panic

Comment (TCeason) [reply src/query/sql/src/planner/binder/setting.rs]: Fixed.

**Review by andylokandy — APPROVED** [review]

File: src/query/sql/src/planner/binder/setting.rs [hunk 6]
```diff
@@ -34,26 +43,39 @@ impl<'a> Binder {
...
+            argument: Box::new(scalar),
+            from_type: Box::new(data_type),
+            target_type: Box::new(DataType::String),
+        });
+        let builder = PhysicalScalarBuilder::new(&schema);
+        let scalar = builder.build(&scalar)?;
+        let expr = scalar.as_expr()?;
+        let block = DataBlock::empty();
+        let func_ctx = self.ctx.try_get_function_context()?;
+        let evaluator = Evaluator::new(&block, func_ctx, &BUILTIN_FUNCTIONS);
```
Comment (sundy-li) [inline src/query/sql/src/planner/binder/setting.rs:69]: We should use `ConstantFolder` than evaluator. Some functions may not have determined value like `rand()`

**Comment by BohuTANG** [comment]
@mergify update

**Pull request merged**